pypesto/hierarchical/optimal_scaling/solver.py

```diff
@@ -987,3 +1045,204 @@ def save_inner_parameters_to_inner_problem(
         inner_parameter.value = x_upper[inner_parameter.category - 1]
     for inner_parameter in problem.get_cat_lb_parameters_for_group(group):
         inner_parameter.value = x_lower[inner_parameter.category - 1]
+
+
+def calculate_censored_obj(
+    category_upper_bounds: List[OptimalScalingParameter],
+    category_lower_bounds: List[OptimalScalingParameter],
+    sim: List[np.ndarray],
+    sigma: List[np.ndarray],
+    quantitative_data: np.ndarray,
+    quantitative_ixs: List[np.ndarray],
+) -> Dict:
+    """Calculate objective function for a group with censored data.
+
+    Parameters
+    ----------
+    category_upper_bounds:
+        The upper bounds for the categories.
+    category_lower_bounds:
+        The lower bounds for the categories.
+    sim:
+        The model simulation.
+    sigma:
+        The noise parameters from AMICI.
+    observable_ids:
+        The observable ids from the model.
+
+    Returns
+    -------
+    Dictionary with the objective function value, dummy success
+    and censoring category bounds.
+    """
+    cat_lb_values = np.array([x.value for x in category_lower_bounds])
+    cat_ub_values = np.array([x.value for x in category_upper_bounds])
+
+    obj = 0
+    # Calculate the objective function for censored data.
+    for cat_lb, cat_ub, cat_ub_par in zip(
+        cat_lb_values,
+        cat_ub_values,
+        category_upper_bounds,
+    ):
+        for sim_i, sigma_i, mask_i in zip(sim, sigma, cat_ub_par.ixs):
+            y_sim = sim_i[mask_i]
+            if len(y_sim) == 0:
+                continue
+            sigma_for_observable = sigma_i[mask_i][0]
+            for y_sim_i in y_sim:
+                if cat_lb > y_sim_i:
+                    y_surrogate = cat_lb
+                elif y_sim_i > cat_ub:
+                    y_surrogate = cat_ub
+                else:
+                    y_surrogate = y_sim_i
+                obj += 0.5 * (
+                    np.log(2 * np.pi * sigma_for_observable**2)
+                    + (y_surrogate - y_sim_i) ** 2 / sigma_for_observable**2
+                )
```

Comment:
Yeah, it currently makes sense only with nllh (even using any ordinal, censored or nonlinear monotone data, because inner_calculator_collector calculated nllh for the other quantitative measurements on default without any other custom settings). 

I'm not checking for custom objectives. But I guess it would make sense to add this only when we add all other PEtab checks we'll need to add for qualitative data specification, which would make sense to make in PEtab, as the other validations are there too. 

Or do you think this should be checked in pypesto because it's specific to qualitative data in pypesto and implemented objectives?

Reply:
> Or do you think this should be checked in pypesto because it's specific to qualitative data in pypesto and implemented objectives?

Yes, because it looks like a pyPESTO limitation, rather than a PEtab limitation. However, if the limitation will be on the PEtab side, then it's fine to leave the validation there. In the case of hierarchical optimization for continuous data, we only support a restricted type of objective function (e.g. no observable transformations, snippet below) in pyPESTO. However, it would be possible to implement other types of objective functions -- so this is a limitation of pyPESTO rather than the PEtab extension. Hence, there is a check for this in pyPESTO.

https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/hierarchical/petab.py#L488-L498

Reply:
Yes, e.g. the noise model and priors on parameters.

Are the assumptions in the ordinal/censored hierarchical problems a superset of the assumptions for continuous data? If so, then fine to add additional checks that are only activated if the additional flag is used. Maybe the checks could even be made observable-dependent, given your inner calculator collector -- this might require a significant refactor, suitable for a different PR.

If they are not a superset, then observable-dependent checks might be necessary, to ensure users do not see errors e.g. unique to the continuous setting, when only using ordinal hierarchical optimization.

Reply:
Yes, I think it'll have to be done observable-dependent. 

Currently, censored and nonlinear-monotone data require additive normal noise. And all of them require continuous data and disallow priors. So it's pretty restrictive. 

It'll take a bit of thought to make it nice. So probably better in a separate PR.